File: raindrop_authserver.py
# Small web server for handling OAuth redirections from Raindrop.io,
# and thereby making OAuth authentication for a local script possible.
# Compatible with both Python 2.7 and 3.X to be somewhat future proof.
# 
# By Andreas Westerlind in 2020
# 

try:
  # Python 3
  from socketserver import TCPServer as HTTPServer
  from http.server import SimpleHTTPRequestHandler
  from urllib.parse import urlparse
  from urllib.parse import parse_qs
except ImportError:
  # Python 2
  from BaseHTTPServer import HTTPServer
  from SimpleHTTPServer import SimpleHTTPRequestHandler
  from urlparse import urlparse
  from urlparse import parse_qs
# Python 2 & 3
import os
import json

# Address and port to use for the webserver. Bind it to 127.0.0.1 so it's not externally reachable,
# and so that we don't get any uneccesary firewall allow/deny dialog
bind_address = ("127.0.0.1", 11038)

# Use this to kill the webserver, as it seems to be wierdly hard to reliably tell it to stop in a better way.
kill_command = "kill $(ps aux|grep raindrop_authserver.py|grep -v 'grep'|awk '{print $2}') 2> /dev/null"

# Define how to handle requests
class AuthRedirectionHandler(SimpleHTTPRequestHandler):
  def do_GET(self):
    try:
      # Get code from Raindrop
      code = parse_qs(urlparse(self.path).query).get("code", str)[0]
    except TypeError:
      # If a request was made without a code, just display an error message
      self.path = "/auth_error.html"
    else:
      # We got the code, so we use that to request a token from Raindrop, which we use a PHP script to do
      # because of issues with SSL root certificates in some versions of Python on macOS,
      # which makes this unreliable to do in Python if we should keep this thing small and simple but still safe.
      token_result = os.popen("/usr/bin/php raindrop-get-token.php '" + code.replace("'", "'\\''") + "'").read()
      if token_result == "success":
        # If we succeeded in getting the token, display information about the authentication being successful
        self.path = "/auth_info.html"
      else:
        # If we failed to get the token, display information about the authentication having failed
        self.path = "/auth_error.html"
    # Stop the webserver in 2 seconds, so we have time to display the info web page before it quits
    os.system("sleep 2 && " + kill_command + " &")
    return SimpleHTTPRequestHandler.do_GET(self)

# Start a countdown for killing the webserver after 20 minutes, so we don't accidentally leave it running forever
os.system("sleep 1200 && " + kill_command + " &")

# Start the web server
server = HTTPServer(bind_address, AuthRedirectionHandler)
server.serve_forever()

File: raindrop_authserver.go
/*
	Function for launching a small web server that handles OAuth redirections from Raindrop.io,
	and thereby making OAuth authentication for a local Raindrop.io client possible.

	By Andreas Westerlind in 2021
*/

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"time"
)

func authserver() {
	pid := fmt.Sprint(os.Getpid())

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		code := ""
		querystring, _ := url.ParseQuery(r.URL.RawQuery)
		if len(querystring["code"]) > 0 {
			code = querystring["code"][0]
			if ok, error_title, error_details := request_token(code); ok {
				http.ServeFile(w, r, "auth_info.html")
			} else {
				auth_error_file, _ := os.ReadFile("auth_error.html")
				auth_error := strings.Replace(strings.Replace(string(auth_error_file), "$ERROR", error_title, 1), "$DETAILS", error_details, 1)
				http.ServeContent(w, r, "auth_error", time.Now(), bytes.NewReader([]byte(auth_error)))
			}
		} else {
			auth_error_file, _ := os.ReadFile("auth_error.html")
			auth_error := strings.Replace(strings.Replace(string(auth_error_file), "$ERROR", "No code variable provided", 1), "$DETAILS", "", 1)
			http.ServeContent(w, r, "auth_error", time.Now(), bytes.NewReader([]byte(auth_error)))
		}
		endcmd := exec.Command("/bin/sh", "-c", "sleep 3 && pkill -P "+pid+" && kill "+pid, "&")
		endcmd.Start()
	})
	endcmd := exec.Command("/bin/sh", "-c", "sleep 1200 && kill "+pid, "&")
	endcmd.Start()
	http.ListenAndServe("127.0.0.1:11038", nil)
}

func request_token(code string) (bool, string, string) {
	// Prepare POST variables
	post_variables := url.Values{}
	post_variables.Add("code", code)
	post_variables.Add("client_id", client_id)
	post_variables.Add("client_secret", client_secret)
	post_variables.Add("redirect_uri", "http://127.0.0.1:11038")
	post_variables.Add("grant_type", "authorization_code")

	resp, err := http.PostForm("https://raindrop.io/oauth/access_token", post_variables)
	if err != nil {
		return false, "Failed to make request to Raindrop.io", err.Error()
	}

	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return false, "Failed to read response from Raindrop.io", err.Error()
	}

	token := RaindropToken{}
	json.Unmarshal([]byte(body), &token)

	if token.Error != "" {
		return false, "Failed to authenticate", token.Error
	}

	location, _ := time.LoadLocation("UTC")
	token.CreationTime = time.Now().In(location).Format("2006-01-02 15:04:05")

	token_json, _ := json.Marshal(token)

	// Save to Keychain
	if err := wf.Keychain.Set("raindrop_token", string(token_json)); err != nil {
		return false, "Failed to save token to Keychain", err.Error()
	}

	return true, "", ""
}
